### PR TITLE
Unroll multiply and add instructions in dotProd_32f

### PR DESCRIPTION
**Pull request by ChipKerchner**
Unroll multiply and add instructions (absorb latencies) into separate accumulators in dotProd_32f - 35% faster.

### CODE REVIEW

**Comment by alalek** [comment]
@terfendail Could you collect performance changes on IA?

File: modules/core/src/matmul.simd.hpp [hunk 1]
```diff
@@ -2511,6 +2511,27 @@ double dotProd_32f(const float* src1, const float* src2, int len)
...
+                              vx_load(src2 + j), v_sum);
+            v_sum1 = v_muladd(vx_load(src1 + j + cWidth),
+                              vx_load(src2 + j + cWidth), v_sum1);
+            v_sum2 = v_muladd(vx_load(src1 + j + (cWidth * 2)),
+                              vx_load(src2 + j + (cWidth * 2)), v_sum2);
+            v_sum3 = v_muladd(vx_load(src1 + j + (cWidth * 3)),
+                              vx_load(src2 + j + (cWidth * 3)), v_sum3);
+        }
+
+        r += v_reduce_sum(v_sum1) + v_reduce_sum(v_sum2) + v_reduce_sum(v_sum3);
```
Comment (tomoaki0705) [inline modules/core/src/matmul.simd.hpp]: ```v_reduce_sum``` is not really fast especially if the HW doesn't have a horizontal sum up instruction.
It's better to reduce the number of call to ```v_reduce_sum```
```
r += v_reduce_sum(v_sum1 + v_sum2 + v_sum3)
```

Comment (alalek) [reply modules/core/src/matmul.simd.hpp]: I believe we can eliminate `v_reduce_sum()` call from here completely:
```
v_sum += v_sum1 + v_sum2 + v_sum3;
```

**Comment by terfendail** [comment]
<details>                                                                                                                                           [58/1943]
<summary>Performance for SSE2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|dot::MatType_Length::(8UC1, 32)|0.000|0.000|1.00|
|dot::MatType_Length::(8UC1, 64)|0.000|0.000|1.00|
|dot::MatType_Length::(8UC1, 128)|0.001|0.001|1.00|
|dot::MatType_Length::(8UC1, 256)|0.004|0.004|1.01|
|dot::MatType_Length::(8UC1, 512)|0.018|0.018|0.97|
|dot::MatType_Length::(8UC1, 1024)|0.082|0.083|0.98|
|dot::MatType_Length::(32SC1, 32)|0.001|0.001|1.03|
|dot::MatType_Length::(32SC1, 64)|0.002|0.002|1.00|
|dot::MatType_Length::(32SC1, 128)|0.008|0.008|1.00|
|dot::MatType_Length::(32SC1, 256)|0.033|0.033|1.00|
|dot::MatType_Length::(32SC1, 512)|0.129|0.129|1.00|
|dot::MatType_Length::(32SC1, 1024)|0.535|0.536|1.00|
|dot::MatType_Length::(32FC1, 32)|0.000|0.000|2.21|
|dot::MatType_Length::(32FC1, 64)|0.001|0.001|1.92|
|dot::MatType_Length::(32FC1, 128)|0.004|0.002|1.80|
|dot::MatType_Length::(32FC1, 256)|0.016|0.009|1.78|
|dot::MatType_Length::(32FC1, 512)|0.076|0.077|0.98|
|dot::MatType_Length::(32FC1, 1024)|0.322|0.332|0.97|
</details>
<details>                                                                                                                                           [34/1943]
<summary>Performance for SSE3 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|dot::MatType_Length::(8UC1, 32)|0.000|0.000|0.93|
|dot::MatType_Length::(8UC1, 64)|0.000|0.000|1.13|
|dot::MatType_Length::(8UC1, 128)|0.001|0.001|1.03|
|dot::MatType_Length::(8UC1, 256)|0.004|0.004|1.02|
|dot::MatType_Length::(8UC1, 512)|0.018|0.017|1.03|
|dot::MatType_Length::(8UC1, 1024)|0.083|0.083|1.00|
|dot::MatType_Length::(32SC1, 32)|0.001|0.001|0.99|
|dot::MatType_Length::(32SC1, 64)|0.002|0.002|1.02|
|dot::MatType_Length::(32SC1, 128)|0.008|0.008|1.03|
|dot::MatType_Length::(32SC1, 256)|0.033|0.032|1.03|
|dot::MatType_Length::(32SC1, 512)|0.129|0.129|1.00|
|dot::MatType_Length::(32SC1, 1024)|0.536|0.511|1.05|
|dot::MatType_Length::(32FC1, 32)|0.000|0.000|1.78|
|dot::MatType_Length::(32FC1, 64)|0.001|0.001|1.92|
|dot::MatType_Length::(32FC1, 128)|0.004|0.002|1.85|
|dot::MatType_Length::(32FC1, 256)|0.016|0.009|1.84|
|dot::MatType_Length::(32FC1, 512)|0.078|0.077|1.01|
|dot::MatType_Length::(32FC1, 1024)|0.329|0.330|1.00|
</details>
<details>                                                                                                                                           [10/1943]
<summary>Performance for SSE4_2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|dot::MatType_Length::(8UC1, 32)|0.000|0.000|1.02|
|dot::MatType_Length::(8UC1, 64)|0.000|0.000|1.00|
|dot::MatType_Length::(8UC1, 128)|0.001|0.001|1.00|
|dot::MatType_Length::(8UC1, 256)|0.004|0.004|1.00|
|dot::MatType_Length::(8UC1, 512)|0.017|0.017|1.00|
|dot::MatType_Length::(8UC1, 1024)|0.090|0.081|1.11|
|dot::MatType_Length::(32SC1, 32)|0.001|0.001|1.00|
|dot::MatType_Length::(32SC1, 64)|0.002|0.002|1.00|
|dot::MatType_Length::(32SC1, 128)|0.008|0.008|1.00|
|dot::MatType_Length::(32SC1, 256)|0.033|0.033|1.00|
|dot::MatType_Length::(32SC1, 512)|0.129|0.130|1.00|
|dot::MatType_Length::(32SC1, 1024)|0.541|0.535|1.01|
|dot::MatType_Length::(32FC1, 32)|0.000|0.000|2.21|
|dot::MatType_Length::(32FC1, 64)|0.001|0.001|1.95|
|dot::MatType_Length::(32FC1, 128)|0.004|0.002|1.79|
|dot::MatType_Length::(32FC1, 256)|0.016|0.009|1.79|
|dot::MatType_Length::(32FC1, 512)|0.075|0.073|1.03|
|dot::MatType_Length::(32FC1, 1024)|0.340|0.331|1.03|
</details>
<details>
<summary>Performance for AVX2 baseline</summary>

| Performance test | Reference time |  PR time | Speedup |
|------------------|----------------|----------|---------|
|dot::MatType_Length::(8UC1, 32)|0.000|0.000|1.00|
|dot::MatType_Length::(8UC1, 64)|0.000|0.000|1.00|
|dot::MatType_Length::(8UC1, 128)|0.001|0.001|1.00|
|dot::MatType_Length::(8UC1, 256)|0.003|0.003|1.00|
|dot::MatType_Length::(8UC1, 512)|0.013|0.013|1.00|
|dot::MatType_Length::(8UC1, 1024)|0.078|0.078|1.01|
|dot::MatType_Length::(32SC1, 32)|0.000|0.000|1.02|
|dot::MatType_Length::(32SC1, 64)|0.002|0.002|1.02|
|dot::MatType_Length::(32SC1, 128)|0.007|0.007|1.01|
|dot::MatType_Length::(32SC1, 256)|0.027|0.027|1.00|
|dot::MatType_Length::(32SC1, 512)|0.109|0.108|1.01|
|dot::MatType_Length::(32SC1, 1024)|0.448|0.453|0.99|
|dot::MatType_Length::(32FC1, 32)|0.000|0.000|1.36|
|dot::MatType_Length::(32FC1, 64)|0.001|0.001|1.18|
|dot::MatType_Length::(32FC1, 128)|0.002|0.002|1.18|
|dot::MatType_Length::(32FC1, 256)|0.009|0.008|1.22|
|dot::MatType_Length::(32FC1, 512)|0.078|0.079|0.99|
|dot::MatType_Length::(32FC1, 1024)|0.326|0.323|1.01|
</details>

**Pull request merged**